[GTK][WPE] WebKitDownload destination should be a path instead of a URI

carlosgcampos opened the pull request:
#### 6fea66849a84ce320661a103dae1c90bcbff5827
<pre>
[GTK][WPE] WebKitDownload destination should be a path instead of a URI
<a href="https://bugs.webkit.org/show_bug.cgi?id=253146">https://bugs.webkit.org/show_bug.cgi?id=253146</a>

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

Destination is handled as a path internally, so using URIs in the API
means we have to convert to/from URI all the time. We can&apos;t change this
in the current API, but for simplicity we can allow passing a path,
while still support URIs for backwards compatibility.
WebKiDownload::created-destination always returns a URI in the old API
for compatibilty, but webkit_download_get_destination() only returns a
URI when a URI was set with webkit_download_set_destination(). The new
API only allows to pass a path, and a path is also passed to
WebKiDownload::created-destination signal.

This patch is based on previous patch by Michael Catanzaro.

* Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
(webkitDownloadDecideDestination):
(webkit_download_class_init):
(webkitDownloadDecideDestinationWithSuggestedFilename):
(webkitDownloadDestinationCreated):
(webkit_download_get_destination):
(webkit_download_set_destination):
* Source/WebKit/UIProcess/API/glib/WebKitDownload.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
(testDownloadRemoteFileError):
(testWebViewDownloadURI):
(testPolicyResponseDownload):
(testDownloadEphemeralContext):
(testDownloadDestinationURI):
(testContextMenuDownloadActions):
(testBlobDownload):
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/261002@main">https://commits.webkit.org/261002@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07c95b20b8895d55963fa454ba65e57f998480f6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110233 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19330 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42889 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1653 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119221 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114184 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20792 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10522 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102492 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115977 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98686 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/30350 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12019 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12631 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8652 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17997 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51304 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14440 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4149 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
```diff
@@ -449,52 +456,73 @@ WebKitURIRequest* webkit_download_get_request(WebKitDownload* download)
 
-    return download->priv->destinationURI.data();
+#if !ENABLE(2022_GLIB_API)
+    if (download->priv->destinationURI)
+        return download->priv->destinationURI.get();
+#endif
+    return download->priv->destination.get();
 }
 
 /**
  * webkit_download_set_destination:
  * @download: a #WebKitDownload
- * @uri: the destination URI
+ * @destination: the destination
```
I think this is worth trying because it's nice to avoid duplicating the comments, but it is an introspection API break, so in theory the rename could break apps that work today. E.g. any app that does `download.set_destination(uri='file:///foo')` will break. Unfortunate. But we got away with doing this once in the past for blacklist/whitelist -> denylist/allowlist, so... hopefully no apps are doing this? It would be a little weird to specify the name of an argument when calling a function with only one argument.

So up to you. I think it's worth trying, but it is an API break.

mcatanzaro replied:
Actually there is another introspection API break here anyway (added nullable annotation to get_destination) and that one is really needed for correctness, so might as well try this too. _Shrug_.

carlosgcampos replied:
I don't think this will break anything.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
```diff
@@ -449,52 +456,73 @@ WebKitURIRequest* webkit_download_get_request(WebKitDownload* download)
+    g_return_val_if_fail(WEBKIT_IS_DOWNLOAD(download), nullptr);
 
-    return download->priv->destinationURI.data();
+#if !ENABLE(2022_GLIB_API)
+    if (download->priv->destinationURI)
+        return download->priv->destinationURI.get();
+#endif
+    return download->priv->destination.get();
```
OK, this seems almost good, but problem is applications that never call webkit_download_set_destination() might still be expecting a URI here. So to be backwards-compatible, I think we need to invert the logic in the `!ENABLE(2022_GLIB_API)` case: return a path only if a path is set with webkit_download_set_destination(), and return a URI otherwise. OK?

carlosgcampos replied:
We would still receive a URI in that case too, because we set the destinationURI in webkitDownloadDecideDestination

carlosgcampos replied:
I've added a test case to cover this.

mcatanzaro replied:
OK, and I see you also added a test to make sure it works properly. Great.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
```diff
@@ -407,23 +411,26 @@ void webkitDownloadFinished(WebKitDownload* download)
 String webkitDownloadDecideDestinationWithSuggestedFilename(WebKitDownload* download, const CString& suggestedFilename, bool& allowOverwrite)
 {
     if (download->priv->isCancelled)
-        return emptyString();
+        return { };
```
This must be fine since the tests are still passing, but I recently learned that the default String is the null string, which is not the same as the empty string, so I'm going to be a lot more careful about converting emptyString() to `{ }` in the future.

carlosgcampos replied:
When using empty string works, null string should work too, because isEmpty() returns true for null strings. It dosnt' work the other way around.

mcatanzaro reviewed (approved):
I like this better than what I did. Thanks!

aperezdc reviewed (approved):
r=me, just a very minor nit in the commit log message

carlosgcampos committed:
[GTK][WPE] WebKitDownload destination should be a path instead of a URI
https://bugs.webkit.org/show_bug.cgi?id=253146

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

Destination is handled as a path internally, so using URIs in the API
means we have to convert to/from URI all the time. We can't change this
in the current API, but for simplicity we can allow passing a path,
while still support URIs for backwards compatibility.
WebKiDownload::created-destination always returns a URI in the old API
for compatibilty, but webkit_download_get_destination() only returns a
URI when a URI was set with webkit_download_set_destination(). The new
API only allows to pass a path, and a path is also passed to
WebKiDownload::created-destination signal.

This patch is based on previous patch by Michael Catanzaro.

* Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
(webkitDownloadDecideDestination):
(webkit_download_class_init):
(webkitDownloadDecideDestinationWithSuggestedFilename):
(webkitDownloadDestinationCreated):
(webkit_download_get_destination):
(webkit_download_set_destination):
* Source/WebKit/UIProcess/API/glib/WebKitDownload.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
(testDownloadRemoteFileError):
(testWebViewDownloadURI):
(testPolicyResponseDownload):
(testDownloadEphemeralContext):
(testDownloadDestinationURI):
(testContextMenuDownloadActions):
(testBlobDownload):
(beforeAll):

Canonical link: https://commits.webkit.org/261002@main
